salt/states/archive.py

```diff
@@ -978,21 +989,36 @@ def extracted(name,
         else:
             cached = __salt__['cp.is_cached'](source_match, saltenv=__env__)
 
+        # If file was already cached we can just re-generate its hash.
         if cached:
             existing_cached_source_sum = _read_cached_checksum(cached)
             log.debug('Existing source sum is: "%s". Expected source sum is "%s"',
                       existing_cached_source_sum, source_sum)
-            if source_sum and existing_cached_source_sum:
-                if existing_cached_source_sum['hsum'] == source_sum['hsum']:
-                    ret['result'] = None if __opts__['test'] else True
-                    ret['comment'] = (
-                        'Archive {0} existing source sum is the same as the '
-                        'expected one and skip_files_list_verify argument was set '
-                        'to True. Extraction is not needed'.format(
-                            salt.utils.url.redact_http_basic_auth(source_match)
-                        )
+
+        # If file was not cached we still could have a pre-existing hash file which would be
+        # generated if update_source was set to True.
+        else:
+            parsed = _urlparse(source_match)
+            # This path would be generated if this file would be cached by file.cached state
+            # We have to mimic it due to questionable logic in the _checksum_file_path function
+            expected_cached_path = salt.utils.path.join(__opts__['cachedir'],
```

Comment:
I don't like it, but I alternative was to change the approach for this whole caching thing in the state which sounds scary for me. Right now it's a bit magical like this archive cache **extrn_files** is only exists in a single line in the state and no other states or modules aware of it. So I have to mimic that logic to generate the expected path to the file.